llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,443 @@
+}
+
+SmallSet<MCPhysReg, 16>
+AIEWawRegRewriter::getBBForbiddenReplacements(const MachineBasicBlock *MBB) {
+  SmallSet<MCPhysReg, 16> BlockedPhysRegs;
+  MCPhysReg PhysReg;
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+
+      PhysReg = Op.getReg();
+      if (Op.getReg().isVirtual())
+        PhysReg = VRM->getPhys(Op.getReg());
+
+      addAliasRegs(BlockedPhysRegs, PhysReg);
+    }
+  }
+
+  LLVM_DEBUG(dbgs() << "Forbidden Physical Register of the bb."
+                    << MBB->getNumber() << "\n");
+  for (auto &Reg : BlockedPhysRegs) {
+    LLVM_DEBUG(dbgs() << printReg(Reg, TRI, 0, MRI) << "\n");
+  }
+
+  return BlockedPhysRegs;
+}
+
+bool AIEWawRegRewriter::replaceReg(const Register Reg,
+                                   SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs) {
+  LLVM_DEBUG(dbgs() << " WAW RegRewriter: Register to replace "
+                    << TRI->getName(VRM->getPhys(Reg)) << "\n");
+  SmallVector<MCPhysReg, 16> Replacements =
+      getPriorityReplacementRegs(Reg, ForbiddenPhysRegs);
+
+  for (const MCPhysReg &PhysReg : Replacements) {
+
+    const LiveInterval &LI = LIS->getInterval(Reg);
+    LiveRegMatrix::InterferenceKind IK = LRM->checkInterference(LI, PhysReg);
+
+    if (IK == LiveRegMatrix::IK_Free) {
+      LLVM_DEBUG(dbgs() << "     WAW Replacement: Virtual Register "
+                        << Register::virtReg2Index(Reg) << " will replace "
+                        << printReg(VRM->getPhys(Reg), TRI, 0, MRI) << " with "
+                        << printReg(PhysReg, TRI, 0, MRI) << '\n');
+      LRM->unassign(LI);
+      LRM->assign(LI, PhysReg);
+      addAliasRegs(ForbiddenPhysRegs, PhysReg);
+      return true;
+    }
+  }
+  return false;
+}
+
+SmallVector<MCPhysReg, 16> AIEWawRegRewriter::getPriorityReplacementRegs(
+    const Register Reg, const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs) {
+  SmallVector<MCPhysReg, 16> PhysRegs;
+  const TargetRegisterClass *RC = MRI->getRegClass(Reg);
+
+  for (const MCPhysReg &PhysReg : RC->getRegisters()) {
+    // iterate Regs of RC and allocate first found reg that does not
+    // interfere
+    if (find(ForbiddenPhysRegs, PhysReg) != ForbiddenPhysRegs.end())
+      continue;
+
+    LiveRegMatrix::InterferenceKind IK =
+        LRM->checkInterference(LIS->getInterval(Reg), PhysReg);
+    if (IK == LiveRegMatrix::IK_Free)
+      PhysRegs.push_back(PhysReg);
+  }
+
+  LLVM_DEBUG(dbgs() << "Renaming Candidate " << printReg(Reg, TRI, 0, MRI)
+                    << " Ordering";
+             for (const MCPhysReg &PhysReg
+                  : PhysRegs) {
+               dbgs() << " " << printReg(PhysReg, TRI, 0, MRI);
+             } dbgs()
+             << "\n";);
+
+  return PhysRegs;
+}
+
+bool AIEWawRegRewriter::isIdentityCopy(const MachineInstr &MI) {
+  if (!MI.isCopy())
+    return false;
+
+  Register DstPhyReg = MI.getOperand(0).getReg();
```

Comment:
nit: maybe a lambda `GetPhyReg`.

like `return GetPhyReg(DstPhyReg, 0) == GetPhyReg(SrcPhyReg, 1);`

Reply:
wrote a function to extract the Physical register from the `MachineOperand`. I kept the variable names, to highlight the functionality.